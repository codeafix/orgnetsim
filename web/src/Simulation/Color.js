const Color = {
    nameArray: [
        "Grey",
        "Blue",
        "Red",
        "Green",
        "Yellow",
        "Orange",
        "Purple",
    ],

    cssColorArray: [
        "SlateGray",
        "Blue",
        "Red",
        "LawnGreen",
        "Yellow",
        "Orange",
        "Purple"
    ],
 
    colorValArray: [
        0,
        1,
        2,
        3,
        4,
        5,
        6
    ],
    
    //return the color name used in the UI for the given color enum
    colorFromVal: function(color) {
        if (color >= 0 && color < this.nameArray.length){
            return this.nameArray[color];
        }
        return "Invalid Color";
    },
    //return the color style name used to fill objects in a d3 chart
    cssColorFromVal: function(color) {
        if (color >= 0 && color < this.nameArray.length){
            return this.cssColorArray[color];
        }
        return this.cssColorArray[0];
    },

    //return a slice of the UI color names
    colorNameSlice: function(maxColors) {
        return this.nameArray.slice(0,maxColors);
    },

    //return a slice of the UI color values
    cssValSlice: function(maxColors) {
        return this.cssColorArray.slice(0,maxColors);
    },

    //return a slice of the UI color values
    colorValSlice: function(maxColors) {
        return this.colorValArray.slice(0,maxColors);
    }
}

export default Color