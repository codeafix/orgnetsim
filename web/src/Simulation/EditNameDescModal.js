import React, { useState, useEffect } from 'react';
import {Modal} from 'react-bootstrap'
import {InputGroup} from 'react-bootstrap'
import {FormControl} from 'react-bootstrap'
import {Button} from 'react-bootstrap'

const EditNameDescModal = (props) => {
    const [sim, setsim] = useState(props.sim);
    const [simname, setsimname] = useState("");
    const [simdescription, setsimdescription] = useState("");

    const handleClose = () => {
        setsimname("");
        setsimdescription("");
        props.closeFunc();
    };

    useEffect(() => {
        if(!props.sim){
            setsimname("");
            setsimdescription("");
        }else{
            setsimname(props.sim.name);
            setsimdescription(props.sim.description);
        }
      },[props.sim]);

    const saveSimulation = () => {
        var sim = props.sim || {}
        sim.name = simname;
        sim.description = simdescription;
        props.saveFunc(sim);
        handleClose();
    }

    return(
        <Modal
            show={props.show}
            onHide={handleClose}
            backdrop="static"
            keyboard={false}
        >
            <Modal.Header closeButton>
                <Modal.Title>Simulation</Modal.Title>
            </Modal.Header>
            <Modal.Body>
                <InputGroup className="mb-3">
                    <InputGroup.Prepend>
                        <InputGroup.Text id="basic-on1">Name</InputGroup.Text>
                    </InputGroup.Prepend>
                    <FormControl
                    placeholder="Simulation name"
                    aria-label="Simulation name"
                    aria-describedby="basic-on1"
                    value={simname}
                    onChange={e => setsimname(e.target.value)}
                    />
                </InputGroup>
                <InputGroup>
                    <InputGroup.Prepend>
                        <InputGroup.Text>Description</InputGroup.Text>
                    </InputGroup.Prepend>
                    <FormControl as="textarea" aria-label="Description" value={simdescription} onChange={e => setsimdescription(e.target.value)}/>
                </InputGroup>
            </Modal.Body>
            <Modal.Footer>
                <Button variant="primary" onClick={saveSimulation}>Save</Button>
                <Button variant="secondary" onClick={handleClose}>Cancel</Button>
            </Modal.Footer>
        </Modal>
    )
}

export default EditNameDescModal