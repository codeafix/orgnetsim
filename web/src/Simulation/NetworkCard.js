import React, { useState, useEffect } from 'react';
import {Card, Form, Modal, FormControl, Button} from 'react-bootstrap'
import API from '../api';
import NetworkGraph from './NetworkGraph';

const NetworkCard = (props) => {
    const [idcol, setidcol] = useState(0);
    const [pcol, setpcol] = useState(2);
    const [delim, setdelim] = useState(",");
    const [filetoupload, setfiletoupload] = useState();
    const [showimpmodal, setshowimpmodal] = useState(false);
    const [hasstep, sethasstep] = useState(false);

    useEffect(() => {
        sethasstep((props.steps || []).length > 0)
    },[props.sim, props.steps]);

    const handleimpclose = () => {
        setshowimpmodal(false);
        setidcol(0);
        setpcol(2);
        setdelim(",");
        setfiletoupload();
    };
    
    const handleimport = () => {
        const fr = new FileReader();

        fr.readAsText(filetoupload);
        fr.onload = function() {
            const base64data = btoa(fr.result);
        
            const pdata = {
                "identifier": idcol,
                "parent": pcol,
                "regex": {
                    "0":"^\\d+_(.*)$",
                    "2":"^\\d+_(.*)$",
                    "3":"\\S+"
                  },
                "delimiter": delim,
                "Payload": base64data
            };
            API.parse(props.sim, pdata).then(response => {
                props.readsim(response.parent);
            });
        };
        handleimpclose();
    };
    
    return(
        <Card className="mb-2 mx-n2">
            <Card.Header><Card.Title>Network<Button size="sm" className="btn btn-primary float-right" onClick={() => setshowimpmodal(true)} disabled={hasstep}>Import</Button></Card.Title></Card.Header>
            <Card.Body>
                <NetworkGraph sim={props.sim} steps={props.steps}/>
            </Card.Body>
            <Modal
                show={showimpmodal}
                onHide={handleimpclose}
                backdrop="static"
                keyboard={false}
            >
                <Modal.Header closeButton>
                    <Modal.Title>Import Network</Modal.Title>
                </Modal.Header>
                <Modal.Body>
                    <Form.Group controlId="form-file">
                        <Form.File label="Select File" onChange={(e) => {
                            if(e.target.files.length) setfiletoupload(e.target.files[0]);
                        }}/>
                    </Form.Group>
                    <Form.Group controlId="form-identifier">
                        <Form.Label>Identifier column</Form.Label>
                        <Form.Control type="number" value={idcol} onChange={e => setidcol(e.target.valueAsNumber)}/>
                        <Form.Text className="text-muted">
                            The column that has the unique identifier in it
                        </Form.Text>
                    </Form.Group>
                    <Form.Group controlId="form-parent">
                        <Form.Label>Parent Identifier Column</Form.Label>
                        <Form.Control type="number" value={pcol} onChange={e => setpcol(e.target.valueAsNumber)}/>
                        <Form.Text className="text-muted">
                            The column that has a parent identifier in it to show hierarchy in the network
                        </Form.Text>
                    </Form.Group>
                    <Form.Group className="mb-3" controlId="form-delimiter">
                        <Form.Label>Delimiter</Form.Label>
                        <FormControl type="string" value={delim} onChange={e => setdelim(e.target.value)}/>
                        <Form.Text className="text-muted">
                            The delimiter used to separate columns in rows of the imported file
                        </Form.Text>
                    </Form.Group>
                </Modal.Body>
                <Modal.Footer>
                    <Button variant="success" onClick={handleimport}>Save</Button>
                    <Button variant="secondary" onClick={handleimpclose}>Cancel</Button>
                </Modal.Footer>
            </Modal>
        </Card>
    )
}

export default NetworkCard