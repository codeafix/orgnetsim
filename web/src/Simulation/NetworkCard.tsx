import React, { useState } from 'react';
import {Card, Button} from 'react-bootstrap'
import NetworkGraph from './NetworkGraph';
import ParseSettings from './ParseSettings';
import { SimInfo } from '../API/SimInfo';
import { Step } from '../API/Step';

type NetworkCardProps = {
    sim:SimInfo;
    steps:Array<Step>;
    readsim(id:string): void;
}

const NetworkCard = (props:NetworkCardProps) => {
    const [showimpmodal, setshowimpmodal] = useState<boolean>(false);
    
    return(
        <Card className="mb-2 mx-n2">
            <Card.Header><Card.Title>Network<Button size="sm" className="btn btn-primary float-right" onClick={() => setshowimpmodal(true)}>Import</Button></Card.Title></Card.Header>
            <Card.Body>
                <NetworkGraph sim={props.sim} steps={props.steps}/>
            </Card.Body>
            <ParseSettings sim={props.sim} steps={props.steps} readsim={props.readsim} showimportmodal={showimpmodal} onclose={() => setshowimpmodal(false)}/>
        </Card>
    )
}

export default NetworkCard