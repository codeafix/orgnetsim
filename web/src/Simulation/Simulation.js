import React, { useState, useEffect } from 'react';
import API from '../api';
import {Link} from 'react-router-dom';

const Simulation = (props) => {
    const [sim, setSim] = useState({});
    
    useEffect(() => {
        API.get(props.match.params.id).then(response => {
            setSim(response);
        })
      });

    return (
        <div>
            <h1>{sim.name}</h1>
            <p>{sim.description}</p>
            <Link to='/'>Home</Link>
        </div>
    )
}

export default Simulation