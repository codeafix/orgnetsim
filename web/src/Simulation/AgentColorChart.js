import React, { useEffect, useRef, useState } from 'react';
import { scaleLinear, max, select, stack, axisBottom, axisLeft, area } from 'd3';
import Color from './Color';
import API from '../api';
import Spinner from 'react-bootstrap/Spinner';

const AgentColorChart = (props) => {
    const chart = useRef(null);
    const [loading, setloading] = useState(false);

    const createChart = (results) => {
        if(!results['colors'][0]) return;
        const maxColors = props.sim.options['maxColors'];
        const margin = {top: 10, right: 60, bottom: 20, left: 40},
            vwidth = chart.current.parentElement.offsetWidth,
            cw = vwidth - margin.left - margin.right,
            vheight = Math.round(cw/1.6),
            ch = vheight - margin.top - margin.bottom;
        
        const resize = () => {
            const w = chart.current.parentElement.offsetWidth,
                h = Math.round((w -  margin.left - margin.right)/1.6);
            select(chart.current).attr('width', w)
                .attr('height', h);
        };

        select(window).on(
            'resize.' + select(chart.current.parentElement).attr('id'), 
            resize
        );
        
        const svg = select(chart.current);

        svg.attr('viewBox', `0 0 ${vwidth} ${vheight}`)
            .attr('preserveAspectRatio', 'xMinYMid')
            .call(resize);

        const dataMax = results['colors'][0].reduce((a, b) => a + b, 0);
        const iterations = results['iterations'];
        const chartData = results['colors'];

        var stackedData = stack()
            .keys(Color.colorValSlice(maxColors))
            .value(function(d, key){
                return d[key]
            })
            (chartData)
            
        //X Axis
        var xh = ch + margin.top;
        var xScale = scaleLinear()
            .domain([0, iterations])
            .range([0, cw]);
        svg.append("g")
            .attr("class", "small")
            .attr("transform", "translate(" + margin.left + "," + xh + ")")
            .call(axisBottom(xScale).ticks(10));
        
        //Add Y axis
        var yScale = scaleLinear()
            .domain([0, dataMax])
            .range([ch, 0]);
        svg.append("g")
            .attr("class", "small")
            .attr("transform", "translate(" + margin.left + "," + margin.top + ")")
            .call(axisLeft(yScale).ticks(10));
          
        svg.selectAll("mylayers")
            .data(stackedData)
            .enter()
            .append("path")
            .style("fill", (d) => Color.cssColorFromVal(d.key))
            .attr("transform", "translate(" + margin.left + "," + margin.top + ")")
            .attr("d", area()
                .x(function(d, i) { return xScale(i); })
                .y0(function(d) { return yScale(d[0]); })
                .y1(function(d) { return yScale(d[1]); })
            );
    }

    useEffect(() => {
        if (!props.sim['id']) {
            return
        }
        setloading(true);
        select(chart.current).selectAll("*").remove();
        API.getResults(props.sim).then(results => {
            createChart(results);
            setloading(false);
        }).catch(err => {
            console.error(err);
            setloading(false);
        });
    },[props.sim]);

    return <div>
            {loading && <Spinner animation="border" variant="info" />}
            <svg class="mb-3" ref={chart}/>
        </div>
}

export default AgentColorChart;