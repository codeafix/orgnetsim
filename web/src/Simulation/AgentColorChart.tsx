import React, { useEffect, useRef, useState } from 'react';
import { scaleLinear, select, stack, axisBottom, axisLeft, area } from 'd3';
import Color from './Color';
import API from '../API/api';
import Spinner from 'react-bootstrap/Spinner';
import { SimInfo } from '../API/SimInfo';
import { Results } from '../API/Results';

type AgentColorChartProps = {
    sim:SimInfo;
}

const AgentColorChart = (props:AgentColorChartProps) => {
    const chart = useRef<SVGSVGElement>(null);
    const [loading, setloading] = useState<boolean>(false);

    const createChart = (results:Results) => {
        if(!results['colors'][0]) return;
        if(!chart.current) return;
        var c = chart.current;
        if(!c.parentElement) return;

        const maxColors = props.sim.options['maxColors'];
        const margin = {top: 10, right: 60, bottom: 20, left: 40},
            vwidth = c.parentElement.offsetWidth,
            cw = vwidth - margin.left - margin.right,
            vheight = Math.round(cw/1.6),
            ch = vheight - margin.top - margin.bottom;
        
        const resize = () => {
            const c = chart.current;
            if(!c) return;
            if(!c.parentElement) return;
            const w = c.parentElement.offsetWidth,
                h = Math.round((w -  margin.left - margin.right)/1.6);
            select(c).attr('width', w)
                .attr('height', h);
        };

        select(window).on(
            'resize.' + select(c.parentElement).attr('id'), 
            resize
        );
        
        const svg = select(c);

        svg.attr('viewBox', `0 0 ${vwidth} ${vheight}`)
            .attr('preserveAspectRatio', 'xMinYMid')
            .call(resize);

        const dataMax = results['colors'][0].reduce((a, b) => a + b, 0);
        const iterations = results['iterations'];
        const chartData = results['colors'];

        var stackedData = stack<number[], number>()
            .keys(Color.colorValSlice(maxColors))(chartData);
            
        //X Axis
        var xh = ch + margin.top;
        var xScale = scaleLinear()
            .domain([0, iterations])
            .range([0, cw]);
        svg.append("g")
            .attr("class", "small")
            .attr("transform", "translate(" + margin.left + "," + xh + ")")
            .call(axisBottom(xScale).ticks(10));
        
        //Add Y axis
        var yScale = scaleLinear()
            .domain([0, dataMax])
            .range([ch, 0]);
        svg.append("g")
            .attr("class", "small")
            .attr("transform", "translate(" + margin.left + "," + margin.top + ")")
            .call(axisLeft(yScale).ticks(10));
          
        svg.selectAll("mylayers")
            .data(stackedData)
            .enter()
            .append("path")
            .style("fill", (d) => Color.cssColorFromVal(d.key))
            .attr("transform", "translate(" + margin.left + "," + margin.top + ")")
            .attr("d", area<number[]>()
                .x(function(d, i) { return xScale(i); })
                .y0(function(d) { return yScale(d[0]); })
                .y1(function(d) { return yScale(d[1]); })
            );
    }

    useEffect(() => {
        if (!props.sim['id']) {
            return
        }
        setloading(true);
        select(chart.current).selectAll("*").remove();
        API.getResults(props.sim).then(results => {
            createChart(results);
            setloading(false);
        }).catch(err => {
            console.error(err);
            setloading(false);
        });
    },[props.sim]);

    return <div id="chart-container">
            {loading && <Spinner animation="border" variant="info" />}
            <svg className="mb-3" ref={chart}/>
        </div>
}

export default AgentColorChart;