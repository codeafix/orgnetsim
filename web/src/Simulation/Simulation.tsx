import React, { useState, useEffect } from 'react';
import API from '../api';
import {Link} from 'react-router-dom';
import {ArrowLeftCircle} from 'react-bootstrap-icons';
import {Card, Button, Row, Col, Container} from 'react-bootstrap'
import EditNameDescModal from '../Home/EditNameDescModal'
import NetworkOptionsCard from './NetworkOptionsCard'
import NetworkCard from './NetworkCard'
import StepsCard from './StepsCard'

const Simulation = (props) => {
    const [sim, setSim] = useState({options:{}});
    const [simsteps, setSimsteps] = useState([]);
    const [showsimeditmodal, setShowsimeditmodal] = useState(false);

    useEffect(() => {
        readsim(props.match.params.id);
      }, [props.match.params.id]);

    const handlesimeditshow = () => setShowsimeditmodal(true);
    const handlesimeditclose = () => setShowsimeditmodal(false);

    const readsim = (id) => {
        API.get(id).then(sresp => {
            setSim(sresp);
            API.getSteps(sresp).then(steps => {
                setSimsteps(steps);
            });
        })
    };
    
    const updatesim = (simtosave) => {
        API.update(simtosave).then(response => {
            setSim(response);
        })
    };

    return (
        <Container>
            <h1 class="px-2 py-2 bg-light rounded"><Link className="btn btn-outline-secondary mr-3 mt-n2 mb-2" to='/' role="button"><ArrowLeftCircle className="mt-n1" /></Link>{sim.name}<Button size="sm" className="btn btn-primary mt-2 mr-2 float-right" onClick={handlesimeditshow}>Edit</Button></h1>
            <Container>
                <Row>
                    <Col sm={8}>
                        <NetworkCard sim={sim} steps={simsteps} readsim={readsim}/>
                        <StepsCard sim={sim} steps={simsteps} readsim={readsim}/>
                    </Col>
                    <Col sm={4}>
                        <Card className="mb-2 mx-n2">
                            <Card.Header><Card.Title>Description</Card.Title></Card.Header>
                            <Card.Body><Card.Text>{sim.description}</Card.Text></Card.Body>
                        </Card>
                        <NetworkOptionsCard sim={sim} setsim={setSim}/>
                    </Col>
                </Row>
            </Container>
            <EditNameDescModal sim={sim} show={showsimeditmodal} saveFunc={updatesim} closeFunc={handlesimeditclose}/>
        </Container>
    )
}

export default Simulation