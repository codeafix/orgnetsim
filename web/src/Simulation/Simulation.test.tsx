import React from 'react';
import {render} from '@testing-library/react';
import '../../node_modules/bootstrap/dist/css/bootstrap.css';
import '../index.css';
import { test } from 'vitest'
import { vi } from 'vitest'
import { act } from 'react-dom/test-utils';
import Simulation from './Simulation';
import { BrowserRouter } from 'react-router-dom';

vi.mock('../API/api');

test('renders without crashing', async () => {
    await act(async () => {
        render(
            <BrowserRouter>
                <Simulation match={{
                        params: {
                            id: ''
                        }
                    }}/>
            </BrowserRouter>
        );
    });
});