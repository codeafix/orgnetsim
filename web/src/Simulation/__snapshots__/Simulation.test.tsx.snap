// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="container-fluid"
  >
    <h1
      class="px-2 py-2 bg-light rounded"
    >
      <a
        class="btn btn-outline-secondary mr-3 mt-n2 mb-2"
        href="/"
        role="button"
      >
        <svg
          class="mt-n1"
          fill="currentColor"
          height="1em"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M1 8a7 7 0 1 0 14 0A7 7 0 0 0 1 8m15 0A8 8 0 1 1 0 8a8 8 0 0 1 16 0m-4.5-.5a.5.5 0 0 1 0 1H5.707l2.147 2.146a.5.5 0 0 1-.708.708l-3-3a.5.5 0 0 1 0-.708l3-3a.5.5 0 1 1 .708.708L5.707 7.5z"
            fill-rule="evenodd"
          />
        </svg>
      </a>
      <button
        class="btn btn-primary mt-2 mr-2 float-right btn btn-primary btn-sm"
        type="button"
      >
        Edit
      </button>
    </h1>
    <div
      class="container-fluid"
    >
      <div
        class="row"
      >
        <div
          class="col-sm-8"
        >
          <div
            class="mb-2 mx-n2 card"
          >
            <div
              class="card-header"
            >
              <div
                class="card-title h5"
              >
                Network
                <button
                  class="btn btn-primary float-right btn btn-primary btn-sm"
                  type="button"
                >
                  Import
                </button>
              </div>
            </div>
            <div
              class="card-body"
            >
              <div
                id="graph-container"
              >
                <svg
                  class="mb-3"
                />
                <button
                  class="btn btn-primary float-right btn btn-primary btn-sm"
                  disabled=""
                  type="button"
                >
                  Play
                </button>
                <button
                  class="btn btn-primary float-right btn btn-primary btn-sm"
                  type="button"
                >
                  Layout
                </button>
              </div>
            </div>
          </div>
          <div
            class="mb-2 mx-n2 card"
          >
            <div
              class="card-header"
            >
              <div
                class="card-title h5"
              >
                Steps
                <button
                  class="btn btn-primary float-right btn btn-primary btn-sm"
                  disabled=""
                  type="button"
                >
                  Export Results
                </button>
                <a
                  download="results.csv"
                  href="#/"
                >
                  ...
                </a>
                <button
                  class="btn btn-primary float-right mr-2 btn btn-primary btn-sm"
                  disabled=""
                  type="button"
                >
                  Run
                </button>
              </div>
            </div>
            <div
              class="small card-body"
            >
              <div
                id="chart-container"
              >
                <svg
                  class="mb-3"
                />
              </div>
              <table
                class="mb-n3 table table-sm table-striped table-bordered"
              >
                <thead>
                  <tr>
                    <th>
                      Iterations
                    </th>
                    <th>
                      Conversations
                    </th>
                  </tr>
                </thead>
                <tbody />
              </table>
            </div>
          </div>
        </div>
        <div
          class="col-sm-4"
        >
          <div
            class="mb-2 mx-n2 card"
          >
            <div
              class="card-header"
            >
              <div
                class="card-title h5"
              >
                Description
              </div>
            </div>
            <div
              class="card-body"
            >
              <p
                class="card-text"
              />
            </div>
          </div>
          <div
            class="mb-2 mx-n2 card"
          >
            <div
              class="card-header"
            >
              <div
                class="card-title h5"
              >
                Network Options
                <button
                  class="btn btn-primary float-right btn btn-primary btn-sm"
                  type="button"
                >
                  Edit
                </button>
              </div>
            </div>
            <div
              class="small card-body"
            >
              <fieldset
                disabled=""
              >
                <span>
                  <div
                    class="form-check"
                  >
                    <input
                      class="form-check-input"
                      readonly=""
                      type="checkbox"
                    />
                    <label
                      class="form-check-label"
                      title=""
                    >
                      Use agents with memory
                    </label>
                  </div>
                </span>
                <span>
                  <label
                    class="pt-2 form-label"
                  >
                    Evangelist list
                  </label>
                  <textarea
                    class="form-control form-control-sm"
                    readonly=""
                  />
                </span>
                <span>
                  <label
                    class="pt-2 form-label"
                  >
                    Initial colors
                  </label>
                  <input
                    class="form-control form-control-sm"
                    readonly=""
                    type="text"
                    value=""
                  />
                </span>
                <span>
                  <div
                    class="pt-2 form-check"
                  >
                    <input
                      class="form-check-input"
                      readonly=""
                      type="checkbox"
                    />
                    <label
                      class="form-check-label"
                      title=""
                    >
                      Link team peers
                    </label>
                  </div>
                </span>
                <span>
                  <label
                    class="pt-2 form-label"
                  >
                    Linked team list
                  </label>
                  <textarea
                    class="form-control form-control-sm"
                    readonly=""
                  />
                </span>
                <span>
                  <label
                    class="pt-2 form-label"
                  >
                    Lone evangelist
                  </label>
                  <input
                    class="form-control form-control-sm"
                    readonly=""
                    type="text"
                    value=""
                  />
                </span>
                <span>
                  <label
                    class="pt-2 form-label"
                  >
                    Maximum colors
                  </label>
                  <input
                    class="form-control form-control-sm"
                    readonly=""
                    type="text"
                    value="0"
                  />
                </span>
              </fieldset>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
